### PR TITLE
Added the example selector node

### PR DESCRIPTION
**Pull request by ivan-cukic**
Added:
 - support for collections of ports
 - support for async ports
 - settings report errors when a property is not set due to incompatible types

### CODE REVIEW

**Review by RalphSteinhagen — CHANGES_REQUESTED** [review]
@ivan-cukic, first off, let me just say: **great job**! I sincerely commend you for tackling this complex task. It's clear that it was no small feat, and I genuinely appreciate the skill and determination it took to get it done. :+1:  🎉 
This was one of the final missing puzzle pieces, and seeing it fall into place is truly rewarding.


After my initial read-through, I've gathered a few thoughts:

1. low-level: I'll leave some specific line-by-line comments on the more minute details. 
2. high-level API and Selector block omplementation: before diving into deeper refactoring, I recommend we first address and bring the unit-test to 'production level' quality since other users could and should use this as a reference or even directly copy from it for their own block implementation.

After that I think we could merge this. 

Following that, I'd suggest creating an issue to discuss potential refactoring directions.

Lastly, if you have the bandwidth, could you address some of the low-hanging fruits in the sonarlint issues? It would be a significant improvement in ensuring our code quality remains high.

Again, thanks for the hard work!

File: test/qa_selector.cpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+    constexpr Selector() noexcept : Selector({}) {}
+
+    Selector(std::initializer_list<std::pair<const std::string, pmtv::pmt>> init_parameter) noexcept : node<Selector<T>, SelectorDoc>(init_parameter) {
+        if (empty(init_parameter)) {
+            return;
+        }
+        std::ignore     = this->settings().apply_staged_parameters();
+
+        selectOut.name  = "selectOut";
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_selector.cpp:117]: Will remove it, though it is not really set as far as I see anywhere, but it should be at some point. If nothing else, would be useful for debugging.

File: include/node.hpp [hunk 2]
```diff
@@ -189,7 +189,7 @@ template<typename Derived>
 concept HasProcessOneFunction = traits::node::can_process_one<Derived>;
 
 template<typename Derived>
-concept HasProcessBulkFunction = traits::node::can_process_bulk<Derived>;
+concept HasProcessBulkFunction = /*requires { typename Derived::has_process_bulk; } ||*/ traits::node::can_process_bulk<Derived>;
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply include/node.hpp:192]: Was playing around with allowing the user to explicitly specify 'I support process_bulk'. Can work, but I think better approach is to test for any function called process_bulk, and then error out if it is not properly implemented.

[the remark this answers is not in the document]
Comment (ivan-cukic) [reply include/node.hpp:192]: Removed.

File: test/qa_selector.cpp [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        expect(eq(block._internalMapping[1].second, outputMap[1]));
+    };
+
+    "Selector<T> vis scheduler"_test = [] {
+        auto thread_pool = std::make_shared<fair::thread_pool::BasicThreadPool>("custom pool", fair::thread_pool::CPU_BOUND, 2, 2); // use custom pool to limit number of threads for emscripten
+
+        auto [graph, sources, sinks, selector] = make_graph(10, 3, 3, { { 0, 0 }, { 1, 1 }, { 2, 2 } });
+
+        // fg::scheduler::simple scheduler(std::move(graph), thread_pool);
+        // scheduler.run_and_wait();
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_selector.cpp:304]: Manually executing the work instead of using the scheduler. Will remove.

File: test/qa_selector.cpp [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        // N.B. some corner case-handling to be added:
+        // * one input mapped to multiple outputs -> check for consistent and produce the same min available() for all mapped outputs
+
+        return fg::work_return_status_t::OK;
+    }
+};
+} // namespace gr::blocks::basic
+
+ENABLE_REFLECTION_FOR_TEMPLATE_FULL((typename T), (gr::blocks::basic::Selector<T>), selectOut, inputs, monitorOut, outputs, nInputs, nOutputs, mapIn, mapOut, backPressure);
+static_assert(fg::HasProcessBulkFunction<gr::blocks::basic::Selector<double>>);
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_selector.cpp:234]: Yes, I thought of that as well. And I'd add some basic checks to the macro so that we can verify as much as possible that a block is correctly implemented.

File: test/qa_selector.cpp [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+                    // report error and/or just ignore
+                    fmt::print("{}: selector() - invalid requested input-output pair ({}, {}) not in range ([0, {}],[0, {}])\n", static_cast<void *>(this), mapIn.value[i], mapOut.value[i], nInputs,
+                               nOutputs);
+                }
+            }
+        }
+    }
+
+    fair::graph::work_return_status_t
+    process_bulk(/* fg::ConsumableSpan */ auto  &select, //
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_selector.cpp:153]: We don't want concepts in the user code as that will make it impossible to detect when they have an incompatible process_bulk. Spelling out types might be a good idea.

[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply test/qa_selector.cpp:153]: Wouldn't concepts fail when instantiating at the CRTP level?

File: test/blocklib/core/selector.hpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                throw std::invalid_argument("Input and output map need to have the same number of elements");
+            }
+
+            for (std::size_t i = 0U; i < mapOut.value.size(); ++i) {
+                _internalMapping[mapIn.value[i]].push_back(mapOut.value[i]);
+            }
+        }
+    }
+
+    using select_reader_t  = typename PortIn<std::uint32_t, Async, Optional>::ReaderType;
```
Comment (RalphSteinhagen) [inline test/blocklib/core/selector.hpp:57]: :thinking: I liked the concepts `ConsumableSpan auto` and `ProducableSpan auto&` better because the`ReaderType` and `WriterType` are `circular_buffer<T>` implementation specific and could be replaced by the user....

OK for now, will experiment with it separately.

File: test/blocklib/core/selector.hpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    using monitor_writer_t = typename PortOut<T, Async, Optional>::WriterType;
+    using input_reader_t   = typename PortIn<T, Async>::ReaderType;
+    using output_writer_t  = typename PortOut<T, Async>::WriterType;
+
+    fair::graph::work_return_status_t
+    process_bulk(select_reader_t                      *select, //
+                 const std::vector<input_reader_t *>  &ins,
+                 monitor_writer_t                     *monOut, //
+                 const std::vector<output_writer_t *> &outs) {
+        static_assert(std::is_same_v<std::remove_cvref_t<decltype(select)>, select_reader_t *>);
```
Comment (RalphSteinhagen) [inline test/blocklib/core/selector.hpp:67]: This seems redundant and more unit-test-ish with the explicit declarations above? Remove?

File: test/blocklib/core/selector.hpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                std::ignore = ins[src_port]->consume(0_UZ);
+
+            } else {
+                // make the implicit consume all available behaviour explicit
+                std::ignore = ins[src_port]->consume(ins[src_port]->available());
+            }
+        }
+
+        // N.B. some corner case-handling to be added:
+        // * one input mapped to multiple outputs -> check for consistent and produce the same min available() for all mapped outputs
```
Comment (RalphSteinhagen) [inline test/blocklib/core/selector.hpp:148]: The `monitorOut` logic seems to have disappeared.
The described corner case doesn't seem to be handled (yet).

File: test/qa_selector.cpp [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+        for (std::size_t iterration = 0; iterration < 100; ++iterration) {
+            for (auto *source : sources) {
+                source->work(std::numeric_limits<std::size_t>::max());
+            }
+            selector->work(std::numeric_limits<std::size_t>::max());
+            for (auto *sink : sinks) {
+                sink->work(std::numeric_limits<std::size_t>::max());
+            }
+        }
+    };
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply test/qa_selector.cpp:246]: I like the tests, neat and concise.

If you could add the unit-test cases for the 'back pressure' cases, checking that the unrouted data is building up in the source buffers. 

Also another test for the 'monitorOut' port which connected or unconnected should not affect the other outputs.

I guess that would round-up and complete the tests for now.

File: test/blocklib/core/selector.hpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+using A           = Annotated<T, description, Arguments...>;
+
+using SelectorDoc = Doc<R""(
+@brief basic multiplexing class to route arbitrary inputs to outputs
+)"">;
+
+template<typename T>
+struct Selector : node<Selector<T>, SelectorDoc> {
+    // port definitions
+    PortIn<std::uint32_t, Async, Optional> selectOut;
```
Comment (RalphSteinhagen) [inline test/blocklib/core/selector.hpp:21]: The `selectOut` and `_selectedSrc` control (if in range), which input is routed to the monitorOut port. The inputs should not be affected regardless if data is being monitored or not.

File: test/blocklib/core/selector.hpp [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+// optional shortening
+template<typename T, fair::meta::fixed_string description = "", typename... Arguments>
+using A           = Annotated<T, description, Arguments...>;
+
+using SelectorDoc = Doc<R""(
+@brief basic multiplexing class to route arbitrary inputs to outputs
+)"">;
+
+template<typename T>
+struct Selector : node<Selector<T>, SelectorDoc> {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/selector.hpp:19]: Please add the documentation of and reference to https://wiki.gnuradio.org/index.php/Selector here.

Since we cannot (yet) paste images here, and updated description text and examples for the (in,out) tuple pairs would be helpful.  N.B. just realised that the GR 3.X block doesn't seem to support (at least it's not documented) the optional 'backPressure' behaviour. We may have added some new functionality.

**Review by RalphSteinhagen — APPROVED** [review]
This looks good. There are some minor unit-test cases and some in-code documentation missing.
Once you addressed these and the CI passes, this branch could IMO be squash-merged to main.

Well done! This bodes well.

File: test/blocklib/core/selector.hpp (hunk 8 above)
Comment (ivan-cukic) [reply test/blocklib/core/selector.hpp:148]: monitor out is unified with the rest.

One input to multiple outputs works - there is a test for that, and for many inputs to one output.
